microprofile/telemetry/src/main/java/io/helidon/microprofile/telemetry/WithSpanInterceptor.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.microprofile.telemetry;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.annotations.WithSpan;
+import jakarta.annotation.Priority;
+import jakarta.inject.Inject;
+import jakarta.interceptor.AroundInvoke;
+import jakarta.interceptor.Interceptor;
+import jakarta.interceptor.InvocationContext;
+
+
+/**
+ * Intercept {@code HelidonWithSpan} annotated method and invoke tracer.
+ */
+@Interceptor
+@Priority(Interceptor.Priority.PLATFORM_BEFORE + 7)
+@HelidonWithSpan
+class WithSpanInterceptor {
+    private static final System.Logger LOGGER = System.getLogger(WithSpanInterceptor.class.getName());
+
+    private final Tracer tracer;
+
+    @Inject
+    WithSpanInterceptor(Tracer tracer) {
+        this.tracer = tracer;
+    }
+
+    /**
+     * Span around method invoke.
+     *
+     * @param context Invocation Context
+     * @return Invocation proceed.
+     * @throws Exception
+     */
+    @AroundInvoke
+    public Object interceptSpan(InvocationContext context) throws Exception {
+
+        LOGGER.log(System.Logger.Level.TRACE, () -> "WithSpan intercepted");
+
+
+        WithSpan annotation = context.getMethod().getAnnotation(WithSpan.class);
+
+        Context parentContext = Context.current();
+        Scope scope = null;
+
+        String spanName = annotation.value();
+
+        //Process span name. Should be class name, as well as inner classes.
+        if (spanName.isEmpty() || spanName.equals("")){
+            String className = context.getMethod().getDeclaringClass().getName();
+            if (className.contains("$")){
+                className = className.substring(className.lastIndexOf("$") + 1);
+            }
+            spanName = className + "." + context.getMethod().getName();
+        }
+
+        Span span = tracer.spanBuilder(spanName)
+                .setSpanKind(annotation.kind())
+                .setParent(parentContext)
+                .startSpan();
+
+        scope = span.makeCurrent();
+
+        try {
+            Object result = context.proceed();
+            span.end();
```

Comment:
span.end() is needed even when exception is recorded, if I am not mistaken.
Not sure if `scope.close()` also  ends the span. 
I think the `span.end()` should be in `finally` block or not there at all

Reply:
This will also simplify the block, as you just do
```java
try {
  return context.proceed();
} catch ...
} finally ...
```

Reply:
Simplified